linux/hax_entry.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2018 Kryptos Logic
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice,
+ *      this list of conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of the copyright holder nor the names of its
+ *      contributors may be used to endorse or promote products derived from
+ *      this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/fs.h>
+#include <linux/miscdevice.h>
+
+#include "../include/hax_interface.h"
+
+MODULE_LICENSE("BSD");
+MODULE_AUTHOR("Intel Corporation");
```

Comment:
It's possible to add multiple `MODULE_AUTHOR` statements.

Listing both will probably be the best approach to reflect the authorship of the Linux kernel module (`Kryptos Logic LLC`) and the HAXM core code (`Intel Corporation`).

Reply:
Sounds good to me :-) Let's do that for now, and I'll let you know if Intel Legal has other ideas.

Reply:
I reached out to colleagues with many years of experience in open source development, and got some important advice from a senior Linux developer/maintainer. A point you and I overlooked earlier is:

> Linux kernel modules should usually be GPL not BSD. It’s highly unlikely the module contains no dependencies on the Linux kernel, so the resulting work is almost certainly GPL (which is fine; BSD to GPL is allowed).

Therefore, this file, as well as probably all other Linux-specific source files, needs to use the GPL license. I know that you wanted a BSD-licensed hypervisor for Linux, but would you compromise on a small number of source files?

In case you do, I've been advised that you should:

* Use the Linux/GPL copyright header with your own copyright notices, so it's clear that this is not an Intel module.
* State that this kernel module is derived from Intel HAXM, and reproduce the Intel copyright as required by the BSD license.
* Call `MODULE_AUTHOR()` only once, using your company name and email. There's no need to include Intel here, since the main purpose is to let people know how to contact the maintainer of the module.

Reply:
Sure, I will apply those changes.

After checking documentation for `MODULE_LICENSE`, it seems the closest match to what we want is `"Dual BSD/GPL"` (defined as: "*GNU Public License v2 or BSD license choice*").

Reply:
> After checking documentation for MODULE_LICENSE, it seems the closest match to what we want is "Dual BSD/GPL"

Hmm, that's a good point. I thought we would have to use just "GPL" because of the "Linux kernel modules should usually be GPL not BSD" comment, but apparently there's also this dual license option, which is actually adopted by a lot of in-tree modules, e.g.:

https://github.com/torvalds/linux/blob/master/mm/zsmalloc.c
https://github.com/torvalds/linux/blob/master/drivers/video/fbdev/intelfb/intelfbdrv.c

So I agree with your choice.

> Also, I was under the impression that GPL-compatibility was determined automatically at build-time depending on whether GPL symbols were imported and which MODULE_LICENSE was chosen.

That's another good point. IIUC, you are saying that in order to retain the dual license status, we have to avoid using any function marked as `EXPORT_SYMBOL_GPL`. That's beyond my knowledge, so please feel free to carry on with this approach while I reach out for help.

Reply:
According to the aforementioned Linux expert, even dual BSD/GPL licensed modules don't have to avoid functions marked as `EXPORT_SYMBOL_GPL`. So please don't worry too much if you end up having to use one of them.

Reply:
Great, good to know. Thanks for the information!